posthog/exception_capture.py

```diff
@@ -0,0 +1,54 @@
+import json
+import sys
+import threading
+from typing import TYPE_CHECKING
+
+from posthog.exception_utils import single_exception_from_error_tuple
+
+if TYPE_CHECKING:
+    from posthog.client import Client
+
+
+class ExceptionCapture:
+    # TODO: Some method of avoiding internal SDK exceptions?
+
+    def __init__(self, client: "Client"):
+        self.client = client
+        self.original_excepthook = sys.excepthook
+        sys.excepthook = self.exception_handler
+        threading.excepthook = self.thread_exception_handler
+
+    def exception_handler(self, exc_type, exc_value, exc_traceback):
+        # don't affect default behaviour.
+        self.capture_exception(exc_type, exc_value, exc_traceback)
+        self.original_excepthook(exc_type, exc_value, exc_traceback)
+
+    def thread_exception_handler(self, args):
+        self.capture_exception(args.exc_type, args.exc_value, args.exc_traceback)
+
+    def capture_exception(self, exc_type, exc_value, exc_traceback):
+        # if hasattr(sys, "ps1"):
+        #     # Disable the excepthook for interactive Python shells
+        #     return
+
+        # Format stack trace like sentry
+        # TODO: For now, we don't support exception chaining and groups, just a single top level exception...
+        exception_info = single_exception_from_error_tuple(exc_type, exc_value, exc_traceback)
+        stack_trace = (
+            exception_info["stacktrace"]["frames"]
+            if "stacktrace" in exception_info and exception_info["stacktrace"].get("frames")
+            else None
+        )
+
+        properties = {
+            "$exception_type": exc_type.__name__,
+            "$exception_message": str(exc_value),
+            "$exception_stack_trace_raw": json.dumps(stack_trace),
+            # TODO: Can we somehow get distinct_id from context here? Stateless lib makes this much harder? 😅
+            # '$exception_personURL': f'{self.client.posthog_host}/project/{self.client.token}/person/{self.client.get_distinct_id()}'
```

Comment:
almost want in-memory registered properties
can imagine code that has info useful in an exception is distant from code that throws it
so you could register super properties including distinct id and we can pull properties out to use them like her for the distinct id but arbitrary exception properties seems useful

obvs wouldn't work as well as in a stateless library but nice to have effectively "per request" properties

#ScopeCreep :p

Reply:
Even with these, I think it will be pretty hard for it to be correct (atleast in python land, seems ideal for PHP land where generally the SDK is initialised per request). Consider:

1. There's a single instance of the SDK per thread - the overarching request can pass info across processes/threads, but there's no good way for SDKs to pass info along (for example, error in celery task, fired from an API request - will have different python SDK installations)
2. An SDK singleton across shared resources, in which case we don't know when one context ends and another one begins (example: 2 requests come in, first one sends a capture event, then second one sends a capture event, then first one errors out, and the super properties are mangled)

I can still try out doing the in-memory registered props and see how it works for us, but doesn't _seem_ likely to me that it will be accurate :p 
At the same time, sentry does manage to connect enough useful information though, which seems to be about manually tagging scope in the right place at the right time (context manager wins), but I'd love for it to be automatic, if people are also capturing other event information around the same time. Guess we can explore this, and if it doesn't work, fallback to what sentry does, i.e. pass the onus onto the user.